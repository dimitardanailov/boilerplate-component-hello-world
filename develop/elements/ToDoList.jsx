import React from 'react';
import { Subject } from 'rxjs';

class ToDoList extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			todos: [],
			text: ''
		};

		this.currentInput = new Subject();
		this.addTodo = new Subject();
	}

	componentDidMount() {
		this.currentInput.subscribe((newValue) => {
			this.setState(() => ({
				text: newValue
			}));
		});

		this.addTodo.subscribe(() => {
			let { todos, text } = this.state;
			todos = todos.concat({ text });

			this.setState({ todos, text: '' });
		});
	}

	render() {
		return (
			<div>
				<form onSubmit={() => this.addTodo.next()} action="javascript:">
					<input value={this.state.text} onChange={(e) => this.currentInput.next(e.target.value)} />
					<button type="submit">Add</button>
				</form>

				<ul>
					{this.state.todos.map((todo, index) => (
						<li key={index}>{todo.text}</li>
					))}
				</ul>
			</div>
		);
	}
}

export default ToDoList;
