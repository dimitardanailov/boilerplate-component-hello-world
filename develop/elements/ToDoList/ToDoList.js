import React from 'react';
import { Subject } from 'rxjs';

class ToDoList extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			todos: [],
			text: ''
		};

		this.currentInput = new Subject();
		this.addTodo = new Subject();
	}

	componentDidMount() {
		this.currentInput.subscribe(newValue => {
			this.setState(() => ({
				text: newValue
			}));
		});

		this.addTodo.subscribe(event => {
			event.preventDefault();

			let { todos } = this.state;
			const { text } = this.state;
			const index = todos.length + 1;
			todos = todos.concat({
				text,
				index
			});

			this.setState({ todos, text: '' });
		});
	}

	render() {
		return (
			<div>
				<form onSubmit={e => this.addTodo.next(e)}>
					<input
						value={this.state.text}
						onChange={e => this.currentInput.next(e.target.value)}
					/>
					<button type="submit">Add</button>
				</form>

				<ul>
					{this.state.todos.map(task => <li key={task.index}>{task.text}</li>)}
				</ul>
			</div>
		);
	}
}

export default ToDoList;
