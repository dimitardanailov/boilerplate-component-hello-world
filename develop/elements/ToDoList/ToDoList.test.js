import React from 'react';
import renderer from 'react-test-renderer';
import ToDoList from './ToDoList';

test('ToDoList - toMatchSnapshot', () => {
	const component = renderer.create(<ToDoList />);

	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('ToDoList has two kids', () => {
	const component = renderer.create(<ToDoList />);

	const tree = component.toJSON();

	expect(tree.children).toHaveLength(2);
});
