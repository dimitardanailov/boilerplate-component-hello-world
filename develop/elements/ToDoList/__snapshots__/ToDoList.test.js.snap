// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoList - toMatchSnapshot 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      value=""
    />
    <button
      type="submit"
    >
      Add
    </button>
  </form>
  <ul />
</div>
`;
